### PR TITLE
Initial WITH clause implementation

### PR DESCRIPTION
**Pull request by naithar**
Added a `with` method as discussed in #40 to enable the use of WITH clauses in query.

Currently only `Select` supports `WITH` clauses.

### CODE REVIEW

File: Sources/SwiftKuery/Select.swift [hunk 1]
```diff
@@ -59,8 +59,10 @@ public struct Select: Query {
     /// ON clause can be represented as `Filter` or a `String` containing raw SQL.
```
Comment (irar2) [inline Sources/SwiftKuery/Select.swift]: Please update the year in Copyright

File: Sources/SwiftKuery/Select.swift [hunk 2]
```diff
@@ -59,8 +59,10 @@ public struct Select: Query {
...
     /// USING clause: an array of `Column` elements that have to match in a JOIN query.
     public private (set) var joins = [(join: Join, on: QueryFilterProtocol?, using: [Column]?)]()
+    
+    public internal (set) var with: [WithTable]?
 
-    private var syntaxError = ""
+    internal var syntaxError = ""
```
Comment (irar2) [inline Sources/SwiftKuery/Select.swift]: Please revert, see the comment below in Select+With

File: Sources/SwiftKuery/With/Select+With.swift [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+ */
+
+// MARK: Select
+
+/// Select query extension to enable with clauses.
+extension Select {
```
Comment (irar2) [inline Sources/SwiftKuery/With/Select+With.swift]: Why did you create an extension? 
I think it needs to be in Select itself.

File: Sources/SwiftKuery/With/Select+With.swift [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+    
+    /// Set tables to be used for WITH clause.
+    ///
+    /// - Parameter tables: A list of the `OrderBy` to apply.
```
Comment (irar2) [inline Sources/SwiftKuery/With/Select+With.swift]: Copy-paste error, this a list of tables, not OrderBy

File: Sources/SwiftKuery/With/Select+With.swift [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+    /// - Returns: A new instance of Select with tables for WITH clause.
+    internal func with(_ tables: [WithTable]) -> Select {
```
Comment (irar2) [inline Sources/SwiftKuery/With/Select+With.swift]: Please remove 'internal'

File: Sources/SwiftKuery/With/With.swift [hunk 6]
```diff
@@ -0,0 +1,33 @@
+/**
+ Copyright IBM Corporation 2016, 2017
```
Comment (irar2) [inline Sources/SwiftKuery/With/With.swift]: Please remove 2016

File: Sources/SwiftKuery/With/With.swift [hunk 7]
```diff
@@ -0,0 +1,33 @@
+/**
```
Comment (irar2) [inline Sources/SwiftKuery/With/With.swift]: We don't have subdirectories in Swift-Kuery, and WITH is too small to start one, please remove the subdirectory

File: Sources/SwiftKuery/With/With.swift [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+ */
+
+// MARK: With
+
+/// Create a query with WITH clause.
```
Comment (irar2) [inline Sources/SwiftKuery/With/With.swift]: Create a query with **a** WITH clause

File: Sources/SwiftKuery/With/With.swift [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+///
+/// - Parameter table: A table for WITH clause.
```
Comment (irar2) [inline Sources/SwiftKuery/With/With.swift]: Parameter table: A table for **a** WITH clause.

File: Sources/SwiftKuery/With/With.swift [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+/// - Parameter query: A SELECT query that will use table from WITH clause.
```
Comment (irar2) [inline Sources/SwiftKuery/With/With.swift]: A SELECT query that will use **the**  table from **the** WITH clause

File: Sources/SwiftKuery/With/With.swift [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+public func with(_ table: WithTable, _ query: Select) -> Select {
+    return with([table], query)
+}
+
+/// Create a query with WITH clause.
+///
+/// - Parameter tables: An array of tables for WITH clause.
+/// - Parameter query: A SELECT query that will use table from WITH clause.
+public func with(_ tables: [WithTable], _ query: Select) -> Select {
```
Comment (irar2) [inline Sources/SwiftKuery/With/With.swift]: Likewise

File: Sources/SwiftKuery/With/WithTable.swift [hunk 12]
```diff
@@ -0,0 +1,49 @@
+/**
+ Copyright IBM Corporation 2016, 2017
```
Comment (irar2) [inline Sources/SwiftKuery/With/WithTable.swift]: Likewise, please remove 2016 and subdirectory

File: Sources/SwiftKuery/With/WithTable.swift [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+ */
+
+// MARK: WithTable
+
+/// Subclasses of the WithTable class are describing a table that is used in WITH clauses.
+open class WithTable: Table {
```
Comment (irar2) [inline Sources/SwiftKuery/With/WithTable.swift]: Please rename to AuxiliaryTable as decided in the issue

File: Sources/SwiftKuery/With/WithTable.swift [hunk 14]
```diff
@@ -0,0 +1,49 @@
...
+    ///
+    /// - Parameter query: A query that will be used in WITH clause.
+    public required init(as query: Query) {
+        super.init()
+        self.query = query
+    }
+    
+    /// Initialize an instance of WithTable.
+    public required init() {
+    }
```
Comment (irar2) [inline Sources/SwiftKuery/With/WithTable.swift]: Isn't this redundant?

File: Tests/SwiftKueryTests/TestWith.swift [hunk 15]
```diff
@@ -0,0 +1,72 @@
+/**
+ Copyright IBM Corporation 2016
```
Comment (irar2) [inline Tests/SwiftKueryTests/TestWith.swift]: Please change to 2017

**Comment by naithar** [comment]
@irar2 please review.

**Comment by irar2** [comment]
@naithar Thanks, I will review tomorrow

**Comment by irar2** [comment]
@naithar Your test doesn't work with Postgres:
```swift
var withTable = AuxTable(as: Select(t1.c.as("d"), t1.b.as("f"), from: t1))        
var s = with(withTable, Select(withTable.d, t.a, from: t).join(withTable))
```
generates
```sql
WITH withTable AS (SELECT t1.c AS d, t1.b AS f FROM t1) 
SELECT withTable.d, t.a 
FROM t JOIN withTable
```
and this query returns a Postgres error: "ERROR:  syntax error at end of input."
Adding withTable to the from list works:
```swift
var withTable = AuxTable(as: Select(t1.c.as("d"), t1.b.as("f"), from: t1))        
var s = with(withTable, Select(withTable.d, t.a, from: [t, withTable]))
```
```sql
WITH withTable AS (SELECT t1.c AS d, t1.b AS f FROM t1) 
SELECT withTable.d, t.a 
FROM t, withTable
```

**Comment by naithar** [comment]
@irar2 I'm looking into it.

**Comment by naithar** [comment]
@irar2 made a change in #52

**Pull request merged**